gradio/queueing.py

```diff
@@ -364,15 +363,73 @@ async def call_prediction(self, events: list[Event], batch: bool):
                 if event.data
             ]
             data.batched = True
-        response = await AsyncRequest(
-            method=AsyncRequest.Method.POST,
-            url=f"{self.server_path}api/predict",
-            json=dict(data),
-            headers={"Authorization": f"Bearer {self.access_token}"},
-            cookies={"access-token": token} if token is not None else None,
-            client=self.queue_client,
+
+        ## TODO: extract the following code copied from routes.py into a shared function.
+        import traceback
+
+        from gradio import utils
+        from gradio.exceptions import Error
+        from gradio.helpers import EventData
+
+        api_name = "predict"
+        app = self.app
+        body = data
+        default_response_class = app.router.default_response_class
+
+        fn_index_inferred = route_utils.infer_fn_index(
+            app=app, api_name=api_name, body=body
+        )
+
+        gr_request = route_utils.compile_gr_request(
+            app=app,
+            body=body,
+            fn_index_inferred=fn_index_inferred,
+            username=username,
+            request=None,
         )
-        return response
+
+        session_state, iterators = route_utils.restore_session_state(app=app, body=body)
+
+        fn_index = body.fn_index
+
+        event_id = getattr(body, "event_id", None)
+        raw_input = body.data
+
+        dependency = app.get_blocks().dependencies[fn_index_inferred]
+        target = dependency["targets"][0] if len(dependency["targets"]) else None
+        event_data = EventData(
+            app.get_blocks().blocks.get(target) if target else None,
+            body.event_data,
+        )
+        batch = dependency["batch"]
+        if not (body.batched) and batch:
+            raw_input = [raw_input]
+        try:
+            with utils.MatplotlibBackendMananger():
+                output = await app.get_blocks().process_api(
+                    fn_index=fn_index_inferred,
+                    inputs=raw_input,
+                    request=gr_request,
+                    state=session_state,
+                    iterators=iterators,
+                    event_id=event_id,
+                    event_data=event_data,
+                )
+            iterator = output.pop("iterator", None)
+            if hasattr(body, "session_hash"):
+                app.iterators[body.session_hash][fn_index] = iterator
+            if isinstance(output, Error):
+                output = default_response_class(output)  # Added
+                raise output
```

Comment:
Why are we duplicating all of the logic from inside the `gr.routes.predict()` method instead of just calling the `gr.routes.predict()` method directly?

Reply:
Sorry I just left it undone but it must be extracted to a shared method as you wrote.

Reply:
Thank you!
I gave a try on it playing around the code a bit, but I thought the first approach looked better then. Let me choose the first way for now.
The reasons are:
* Obtaining the endpoint handler from the router object is not documented either in FastAPI or Starlette documents, so it seems not a public API.
* Starlette does it for routing like [this](https://github.com/encode/starlette/blob/a8b8856ce393a82ab4a714131085dbc4d658e34d/starlette/routing.py#L740-L758C19) internally (FastAPI delegates the routing to Starlette). Doing it by ourselves would be reimplementation of the framework internals, and it can be seen the same as [another approach I suggested emulating the ASGI protocol](https://github.com/gradio-app/gradio/pull/5124/commits/67c432dedf240ff07fe5d96a9234daa84d1ec5e0), which we considered not natural, in that they are imitation of framework's roles.